### PR TITLE
zos add --all should only add non-library project contracts

### PR DESCRIPTION
**Pull request by siromivel**
Addresses #754

This fixes the issue where `zos add --all` adds non-project and library contracts. I identified project contracts by recursing through the contracts directory and aggregating an array of contract names stripped of file extensions. We then check against that array of project contracts while going through the built contracts to add and only add those that are found in the project. 

Libraries are excluded by checking the through the contracts' compiled JSON interface, this may be fragile as it appeared on first glance that the `contractKind` always appears on the second element of the `contractData.ast.nodes` array - if that's not the case it would be simple to refactor this to iteratively check for that value.

### CODE REVIEW

**Comment by spalladino** [comment]
This is great, thanks so much @siromivel! @jcarpanelli could you take a look at this, since you are currently working on the commands interface?

**Comment by siromivel** [comment]
I just pushed a quick fix forr #760 to this branch as I agree it makes sense to clean up these issues together - let me know if that's an issue and I can prune that commit.

File: packages/cli/src/models/initializer/truffle/Truffle.ts [hunk 1]
```diff
@@ -72,8 +72,11 @@ const Truffle = {
         .filter((name) => {
           const contract = FileSystem.parseJsonIfExists(`${buildDir}/${name}`);
           if (contract) {
+            let isLibrary = contract.ast ? contract.ast.nodes[1].contractKind === "library" : false;
             const projectDir = buildDir.replace('build/contracts', '');
-            return contract.bytecode.length > 2 && contract.sourcePath.indexOf(projectDir) === 0;
+            return (contract.bytecode.length > 2 &&
+                    contract.sourcePath.indexOf(projectDir) === 0) &&
+                    !isLibrary
```
Comment (siromivel) [inline packages/cli/src/models/initializer/truffle/Truffle.ts]: This addresses #760 - note that it will completely omit libraries from the return of this method, if that's an issue this could be done in a more granular fashion. All tests passed locally and I couldn't immediately see a problem with omitting the libraries.

File: packages/cli/src/models/local/LocalController.ts [hunk 2]
```diff
@@ -57,13 +57,37 @@ export default class LocalController {
   }
 
   public addAll(): void {
+    function trimExtension(file: string) {
+      let fileExtRegex = new RegExp(/\.[^\.]*$/);
+      return file.replace(fileExtRegex, "");
+    }
+
+    function getAllFilenamesAtPath(path: string): string[] {
+      return fs.readDir(path).reduce((contracts: any, child: string) => {
```
[the remark this answers is not in the document]
Comment (siromivel) [reply packages/cli/src/models/local/LocalController.ts]: This function is actually completely unnecessary now that I know relying on `sourcePath` is reasonable; it needed to be recursive originally as it was meant to go through the uncompiled contracts directory and in the test cases was not finding contracts in subdirectorys (specifically the `/contracts/mocks` path where the various Impl contracts live.)

Next commit will remove this monstrosity in favor of the simpler solution :grin:

**Comment by siromivel** [comment]
I'll look at fleshing out the test cases this afternoon and get the PR updated with those!

**Comment by siromivel** [comment]
Just pushed a commit cleaning up the logic for checking if a contract belongs to the project and broke the tests for this out into their own test block. I also added a `WithExternalPackageImplV1` contract to `ImplV1.sol` that relies on the `mock-stdlib` package and assertions ensuring the external package and library aren't being added.

File: packages/cli/src/models/initializer/truffle/Truffle.ts [hunk 3]
```diff
@@ -72,8 +72,12 @@ const Truffle = {
         .filter((name) => {
           const contract = FileSystem.parseJsonIfExists(`${buildDir}/${name}`);
           if (contract) {
+            const isLibrary = contract.ast ? contract.ast.nodes[1].contractKind === "library" : false;
```
Comment (jcarpanelli) [inline packages/cli/src/models/initializer/truffle/Truffle.ts]: I believe this is not generic enough. Image that you have a `Foo.sol` in which you have, some contracts and libraries defined:
```solidity
pragma solidity ^0.5.0;

contract MyContract {}
library MyLibrary {}
library MyOtherLibrary {}
```
In this case, when reading the contract ast, you will find the pragma-related info in the first position and the `MyContract` contract-related info in the second position (i.e., `contract.ast[1]`), and not the info related to the library. You will struggle with the same problem if you have imports before declaring your libraries.
In order to solve this, you can check not only the `contractKind` value but also the `name` to be sure you are picking the right element of the `nodes` array.
For instance:
```ts
const isLibrary = !!contract.ast.nodes.find(node => node.contractKind === 'library' && node.name === contract.contractName);
```
(Please double check that I'm not commiting any mistake with the node's attributes :laughing:)

Comment (jcarpanelli) [reply packages/cli/src/models/initializer/truffle/Truffle.ts]: nit: use single quotes instead of double quotes or the linter will complain.

File: packages/cli/src/models/local/LocalController.ts [hunk 4]
```diff
@@ -57,13 +57,20 @@ export default class LocalController {
...
+    const buildFolder = Contracts.getLocalBuildDir();
+    const sourceFolder = Contracts.getLocalContractsDir();
+
+    fs.readDir(buildFolder).forEach((file) => {
+      const path = `${buildFolder}/${file}`;
       if(this.hasBytecode(path)) {
         const contractData = fs.parseJson(path);
-        const contractName = contractData.contractName;
-        this.add(contractName, contractName);
+        const isLibrary = contractData.ast.nodes[1].contractKind === "library";
```
Comment (jcarpanelli) [inline packages/cli/src/models/local/LocalController.ts]: Same as above regarding how to pick the correct ast node item and regarding the quotation.

File: packages/cli/test/scripts/add.test.js [hunk 5]
```diff
@@ -78,6 +79,17 @@ contract('add script', function() {
     this.packageFile.contract('ImplV2').should.eq('ImplV2')
   })
 
+  it('should not add solidity libraries or external packages when adding all', function() {
```
Comment (jcarpanelli) [inline packages/cli/test/scripts/add.test.js]: nit: `should not add solidity libraries or contracts from external packages when adding all`

**Comment by siromivel** [comment]
Took care of the small stuff in here yesterday; planning to take a deeper look at the nodes data structure and get this finished up this afternoon 🚀

File: packages/cli/src/models/local/LocalController.ts [hunk 6]
```diff
@@ -64,10 +64,10 @@ export default class LocalController {
       const path = `${buildFolder}/${file}`;
       if(this.hasBytecode(path)) {
         const contractData = fs.parseJson(path);
-        const isLibrary = contractData.ast.nodes[1].contractKind === "library";
         const isProjectContract = contractData.sourcePath.indexOf(sourceFolder) === 0;
+        const isLibrary = contractData.ast && contractData.ast.nodes.find((node) => node.name === contractData.contractName && node.contractKind === 'library');
```
Comment (siromivel) [inline packages/cli/src/models/local/LocalController.ts]: The existence check for contractData.ast may be unnecessary - it was an edge case I ran into in the Truffle tests when working with the contract with bytecode `0x`, that contract has no ast on it. I spent a while trying to compose a contract that would compile into the same an was unsuccessful; a completely blank contract file won't compile at all and even just a pragma declaration will generate some viable bytecode. If there's no real world scenario that leads to this I'd suggest possibly modifying the truffle test and removing this existence check.

**Comment by siromivel** [comment]
@jcarpanelli I took care of the feedback this afternoon; an integration test is timing out on Travis but I think it may just be flaky - ran the suite 5 times locally and the same timeout occured on one of those runs.

I was also thinking it may be nice to abstract the logic for checking if a contract is a library so that we can avoid duplicating that code(with this PR it will be used in both `Truffle.js` and `add.test.js`). I poked around the repo a bit and found the various `helpers` and `utilities` folders. It also might make sense to expose it as a property at the `contract` level so that we don't have to traverse the nodes every time we want to check it. Would one of those be an appropriate location for this logic or is there a better place that I'm not familiar with yet? Thanks!
